### PR TITLE
Provide config option to decide the hash-type (hash, full_hash, etc.) used by the db as key. 

### PR DESCRIPTION
**Pull request by koparasy**
The PR provides a configuration option that defines the type of a hash to be used by the spack-DB.  

The config option can be referenced by the projection option and create installation paths that include different hash-types. So, when hash_type='full_hash' spack supports multiple installations of the same package with different dependencies. Moreover, the lock mechanism uses the hash_type value to create the respective locks. Therefore, when hash_type='full_hash' spack supports concurrent installations of the same package with different dependencies. 

To summarize, the lock mechanism, the spack-DB key values and, the directory prefixes should have a mechanism to control their behavior. This branch is an initial effort towards that direction.

### CODE REVIEW

**Comment by alalazo** [comment]
Since #28504 we have a single hash (the full hash), I don't think we plan to go back. Closing and pinging @becker33 and @tgamblin for their information
